core/src/test/scala/akka/persistence/cassandra/journal/RecoveryLoadSpec.scala

```diff
@@ -0,0 +1,152 @@
+      case RecoveryCompleted =>
+        replayEndTime = System.nanoTime()
+    }
+
+    def receiveCommand: Receive = {
+      case Init(numberOfEvents) =>
+        self ! Next(numberOfEvents)
+        context.become(init(sender()))
+      case Delete(seqNr) =>
+        deleteMessages(seqNr)
+      case GetMetrics =>
+        sender() ! Metrics(
+          snapshotDuration = (snapshotEndTime - startTime).nanos,
+          replayDuration1 = (replayStartTime - snapshotEndTime).nanos,
+          replayDuration2 = (replayEndTime - replayStartTime).nanos,
+          replayedEvents,
+          totalDuration = (replayEndTime - startTime).nanos
+        )
+    }
+
+    def init(replyTo: ActorRef): Receive = {
+      case Next(remaining) =>
+        if (remaining == 0)
+          replyTo ! InitDone
+        else {
+          val tags = tagging(lastSequenceNr)
+          val event =
+            if (tags.isEmpty) s"event-$lastSequenceNr"
+            else Tagged(s"event-$lastSequenceNr", tags)
+
+          persist(event) { _ =>
+            if (lastSequenceNr % snapshotEvery == 0) {
+              saveSnapshot(s"snap-$lastSequenceNr")
+            }
+            self ! Next(remaining - 1)
+          }
+        }
+    }
+
+  }
+
+}
+
+class RecoveryLoadSpec extends TestKit(ActorSystem("RecoveryLoadSpec", RecoveryLoadSpec.config))
+  with ImplicitSender with WordSpecLike with Matchers with CassandraLifecycle {
+
+  import RecoveryLoadSpec._
+
+  override def systemName: String = "RecoveryLoadSpec"
+
+  private def printMetrics(metrics: Metrics): Unit = {
+    println(s"  snapshot recovery took ${metrics.snapshotDuration.toMillis} ms")
+    println(s"  replay init took ${metrics.replayDuration1.toMillis} ms")
+    println(s"  replay of ${metrics.replayedEvents} events took ${metrics.replayDuration2.toMillis} ms")
+    println(s"  total recovery took ${metrics.totalDuration.toMillis} ms")
+  }
+
+  "Recovery" should {
+
+    "have some reasonable performance" in {
+      val pid = "a1"
+      val snapshotEvery = 1000
+      val tagging: Long => Set[String] = { _ => Set.empty }
+      //      val tagging: Long => Set[String] = { seqNr =>
+      //        if (seqNr % 10 == 0) Set("blue")
+      //        else if (seqNr % 17 == 0) Set("blue", "green")
+      //        else Set.empty
+      //      }
+      val prps = props(persistenceId = pid, snapshotEvery, tagging)
+
+      val p1 = system.actorOf(prps)
+      p1 ! Init(numberOfEvents = 9999)
+      expectMsg(20.seconds, InitDone)
+      //            p1 ! Delete(seqNr = 2000000)
+      //            Thread.sleep(3000)
+      system.stop(p1)
+
+      // wait > 2 * scanning-flush-interval
+      Thread.sleep(4500)
+
+      (1 to 3).foreach { n =>
+        val p2 = system.actorOf(prps)
+        p2 ! GetMetrics
+        val metrics = expectMsgType[Metrics](3.seconds)
+        println(s"iteration #$n")
+        printMetrics(metrics)
+        system.stop(p2)
+      }
+    }
+  }
```

Comment:
I used this test to verify the performance improvements.
I populated db with the following test data (different pids):
* a1: 10 events, no tags
* a2: 100 events, no tags
* a3: 10k events, no tags, snapshot every 1000, replay 999 events
* a4: 100k events, no tags, snapshot every 1000, replay 999 events
* a5: 1000k events, no tags, snapshot every 1000, replay 999 events
* a6: 100k events, some tags, snapshot every 1000, replay 999 events

The raw tag scanning without using the new tag_scanning table was improved for 1000k events from 18 s to 15 s.

The full recovery time for 1000k events with the new tag_scanning table improved from 18 s to 83 ms.
For 100k events, from 1584 ms to 142 ms.

Here are the full results:

```
a1: 10 events, no tags
a2: 100 events, no tags
a3: 10k events, no tags, snapshot every 1000, replay 999 events
a4: 100k events, no tags, snapshot every 1000, replay 999 events
a5: 1000k events, no tags, snapshot every 1000, replay 999 events
a6: 100k events, some tags, snapshot every 1000, replay 999 events

*** OLD

a1:
  snapshot recovery took 0 ms
  replay init took 19 ms
  replay of 10 events took 1 ms
  total recovery took 20 ms

a2:
  snapshot recovery took 0 ms
  replay init took 10 ms
  replay of 100 events took 11 ms
  total recovery took 21 ms

a3:
  snapshot recovery took 3 ms
  replay init took 22 ms
  replay of 999 events took 87 ms
  total recovery took 113 ms

a4:
  snapshot recovery took 2 ms
  replay init took 1952 ms
  replay of 999 events took 16 ms
  total recovery took 1972 ms

a5:
  snapshot recovery took 4 ms
  replay init took 18011 ms
  replay of 999 events took 36 ms
  total recovery took 18052 ms

a6:
  snapshot recovery took 5 ms
  replay init took 14 ms
  replay of 999 events took 81 ms
  total recovery took 101 ms

*** NEW without tag_scanning

a1:
  snapshot recovery took 0 ms
  replay init took 23 ms
  replay of 10 events took 1 ms
  total recovery took 25 ms

a2:
  snapshot recovery took 0 ms
  replay init took 14 ms
  replay of 100 events took 8 ms
  total recovery took 22 ms

a3:
  snapshot recovery took 2 ms
  replay init took 177 ms
  replay of 999 events took 20 ms
  total recovery took 201 ms

a4:
  snapshot recovery took 2 ms
  replay init took 1565 ms
  replay of 999 events took 16 ms
  total recovery took 1584 ms

a5:
  snapshot recovery took 2 ms
  replay init took 14886 ms
  replay of 999 events took 20 ms
  total recovery took 14909 ms

a6:
  snapshot recovery took 3 ms
  replay init took 1674 ms
  replay of 999 events took 16 ms
  total recovery took 1695 ms

*** NEW with tag_scanning

a1:
  snapshot recovery took 0 ms
  replay init took 15 ms
  replay of 10 events took 1 ms
  total recovery took 17 ms

a2:
  snapshot recovery took 0 ms
  replay init took 18 ms
  replay of 100 events took 12 ms
  total recovery took 30 ms

a3:
  snapshot recovery took 3 ms
  replay init took 57 ms
  replay of 999 events took 34 ms
  total recovery took 94 ms

a4:
  snapshot recovery took 3 ms
  replay init took 12 ms
  replay of 999 events took 127 ms
  total recovery took 142 ms

a5:
  snapshot recovery took 4 ms
  replay init took 19 ms
  replay of 999 events took 59 ms
  total recovery took 83 ms

a6:
  snapshot recovery took 5 ms
  replay init took 14 ms
  replay of 999 events took 68 ms
  total recovery took 89 ms

```

Reply:
Nice